include/cuda/memory_resource

```diff
@@ -0,0 +1,309 @@
+//
+// Part of the CUDA Toolkit, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _CUDA_MEMORY_RESOURCE
+#define _CUDA_MEMORY_RESOURCE
+
+#include <cuda/std/type_traits>
+
+#include <cuda/std/detail/__concepts>
+#include <cuda/std/detail/__config>
+
+#include <cuda/std/detail/__pragma_push>
+
+#if _LIBCUDACXX_STD_VER > 11
+_LIBCUDACXX_BEGIN_NAMESPACE_CUDA
+
+///////////////////////////////////////////////////////////////////////////////
+// memory_resource
+_LIBCUDACXX_TEMPLATE(class _Ret, class _Ptr)
+(requires _CUDA_VSTD::is_same_v<_Ptr, _Ret>) //
+    void _Returns_matching(_Ptr) {}
+
+/// \concept _Has_member_allocate
+/// \brief The \c _Has_member_allocate concept
+template <class _Resource>
+_LIBCUDACXX_CONCEPT_FRAGMENT(
+    _Has_member_allocate_,
+    requires(_Resource &__res, size_t __bytes, size_t __alignment) //
+    (_Returns_matching<void *>(__res.allocate(__bytes, __alignment))));
+template <class _Resource>
+_LIBCUDACXX_CONCEPT _Has_member_allocate =
+    _LIBCUDACXX_FRAGMENT(_Has_member_allocate_, _Resource);
+
+/// \concept _Has_member_deallocate
+/// \brief The \c _Has_member_deallocate concept
+template <class _Resource>
+_LIBCUDACXX_CONCEPT_FRAGMENT(_Has_member_deallocate_,
+                             requires(_Resource &__res, void *__ptr,
+                                      size_t __bytes,
+                                      size_t __alignment) //
+                             (__res.deallocate(__ptr, __bytes, __alignment)));
+template <class _Resource>
+_LIBCUDACXX_CONCEPT _Has_member_deallocate =
+    _LIBCUDACXX_FRAGMENT(_Has_member_deallocate_, _Resource);
+
+/// \concept resource
+/// \brief The \c resource concept
+template <class _Resource>
+_LIBCUDACXX_CONCEPT resource =
+    _Has_member_allocate<_Resource> &&_Has_member_deallocate<_Resource>
+    /* && _CUDA_VSTD::equality_comparable<_Resource> */;
+
+/// \concept has_property
+/// \brief The \c has_property concept
+template <class _Resource, class _Property>
+_LIBCUDACXX_CONCEPT_FRAGMENT(_Has_property_,
+                             requires(const _Resource &__res) //
+                             (get_property(__res, _Property{})));
+template <class _Resource, class _Property>
+_LIBCUDACXX_CONCEPT has_property = _LIBCUDACXX_FRAGMENT(_Has_property_,
+                                                        _Resource, _Property);
+
+/// \concept has_property_with
+/// \brief The \c has_property_with concept
+template <class _Resource, class _Property, class _Return>
+_LIBCUDACXX_CONCEPT_FRAGMENT(
+    _Has_property_with_,
+    requires(const _Resource &__res) //
+    (_Returns_matching<_Return>(get_property(__res, _Property{})),
+     _Returns_matching<_Return>(
```

Comment:
Actually, I would prefer if we could reuse `same_as` or `is_same_v` but I had trouble in my experiments. @ericniebler what is the preferred way to enforce a certain return type?

Reply:
Probably better at this point to abandon the macros and define the concept twice under a compile-time switch: once for real concepts and once for emulate.

Reply:
Now the sad story is that there is no c++20 yet soo, this is all we have 😿

Reply:
Are you saying libcu++ doesn't support C++20?